core/storage/btree.rs

```diff
@@ -2229,30 +2251,74 @@ impl BTreeCursor {
                             cell_idx,
                             self.usable_space() as u16,
                         )?;
-                        !contents.overflow_cells.is_empty()
                     };
                     self.stack.set_cell_index(cell_idx as i32);
-                    if overflow {
-                        // A balance will happen so save the key we were inserting
-                        tracing::debug!(page = page.get().get().id, cell_idx, "balance triggered:");
+                    let write_info = self
+                        .state
+                        .mut_write_info()
+                        .expect("write info should be present");
+                    write_info.state = WriteState::CheckNeedsBalancing { page: page.clone() };
+                    continue;
+                }
+                WriteState::Overwrite {
+                    page,
+                    cell_idx,
+                    mut state,
+                } => {
+                    turso_assert!(
+                        page.get().is_loaded(),
+                        "page {}is not loaded",
+                        page.get().get().id
+                    );
+                    if matches!(
+                        self.overwrite_cell(page.clone(), cell_idx, record, &mut state)?,
+                        IOResult::IO
```

Comment:
This makes me think it could be worth for us to do:
```rust
impl IoResult {
    pub fn is_io(&self) -> bool {
        matches!(self, Self::IO)
    }
}
```
Not something for this pr obviously, just wanted to leave my thought